File: nobodywho/swift/src/Encoder.swift
import Foundation
import NobodyWhoGenerated

/// An encoder for generating text embeddings.
///
/// ```swift
/// let encoder = try await Encoder.fromPath(modelPath: "embedding-model.gguf")
/// let embedding = try await encoder.encode("Hello world")
/// ```
public class Encoder {
    private let inner: NobodyWhoGenerated.RustEncoder

    /// - Parameters:
    ///   - model: A loaded model to use for encoding.
    ///   - contextSize: Maximum context size in tokens. Defaults to 4096 if nil.
    /// - Throws: `NobodyWhoError` if the model is not an embedding model.
    public init(model: Model, contextSize: UInt32? = nil) throws {
        self.inner = try NobodyWhoGenerated.RustEncoder(model: model.inner, contextSize: contextSize)
    }

    /// Create an encoder directly from a model path.
    public static func fromPath(
        modelPath: String,
        useGpu: Bool = true,
        contextSize: UInt32? = nil
    ) async throws -> Encoder {
        let model = try await Model.load(modelPath: modelPath, useGpu: useGpu)
        return try Encoder(model: model, contextSize: contextSize)
    }

    /// Encode text into an embedding vector.
    public func encode(_ text: String) async throws -> [Float] {
        return try await inner.encode(text: text)
    }
}

File: nobodywho/kotlin/src/Encoder.kt
package ai.nobodywho

import java.io.Closeable
import uniffi.nobodywho.RustEncoder as InternalEncoder

/**
 * An embedding encoder for generating text embeddings.
 *
 * ```kotlin
 * val encoder = Encoder.fromPath("embeddings.gguf")
 * val embedding = encoder.encode("Hello world")
 * ```
 */
class Encoder(
    model: Model,
    contextSize: UInt? = null
) : Closeable {
    private val inner: InternalEncoder = InternalEncoder(model.inner, contextSize)

    companion object {
        /** Create an encoder directly from a model path. */
        suspend fun fromPath(
            modelPath: String,
            useGpu: Boolean = true,
            contextSize: UInt? = null
        ): Encoder = Encoder(Model.load(modelPath, useGpu), contextSize)
    }

    suspend fun encode(text: String): List<Float> = inner.encode(text)

    /** Free the underlying Rust resources. */
    fun destroy() = inner.destroy()
    override fun close() { destroy() }
}
